components/api-resource-mgt/org.wso2.carbon.identity.api.resource.mgt/src/main/java/org/wso2/carbon/identity/api/resource/mgt/dao/impl/CacheBackedAPIResourceMgtDAO.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.api.resource.mgt.dao.impl;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.api.resource.mgt.APIResourceMgtException;
+import org.wso2.carbon.identity.api.resource.mgt.APIResourceMgtServerException;
+import org.wso2.carbon.identity.api.resource.mgt.cache.APIResourceCacheById;
+import org.wso2.carbon.identity.api.resource.mgt.cache.APIResourceCacheByIdentifier;
+import org.wso2.carbon.identity.api.resource.mgt.cache.APIResourceCacheEntry;
+import org.wso2.carbon.identity.api.resource.mgt.cache.APIResourceIdCacheKey;
+import org.wso2.carbon.identity.api.resource.mgt.cache.APIResourceIdentifierCacheKey;
+import org.wso2.carbon.identity.api.resource.mgt.dao.APIResourceManagementDAO;
+import org.wso2.carbon.identity.application.common.model.APIResource;
+import org.wso2.carbon.identity.application.common.model.ApplicationBasicInfo;
+import org.wso2.carbon.identity.application.common.model.Scope;
+import org.wso2.carbon.identity.core.model.ExpressionNode;
+
+import java.util.List;
+
+/**
+ * This class implements the {@link APIResourceManagementDAO} interface.
+ */
+public class CacheBackedAPIResourceMgtDAO implements APIResourceManagementDAO {
+
+    private static final Log log = LogFactory.getLog(CacheBackedAPIResourceMgtDAO.class);
+    private APIResourceCacheByIdentifier apiResourceCacheByIdentifier = null;
+    private APIResourceCacheById apiResourceCacheById = null;
+    private final APIResourceManagementDAO apiResourceManagementDAO;
+
+    public CacheBackedAPIResourceMgtDAO(APIResourceManagementDAO apiResourceManagementDAO) {
+
+        this.apiResourceManagementDAO = apiResourceManagementDAO;
+        apiResourceCacheByIdentifier = APIResourceCacheByIdentifier.getInstance();
+        apiResourceCacheById = APIResourceCacheById.getInstance();
+    }
+
+    @Override
+    public List<APIResource> getAPIResources(Integer limit, Integer tenantId, String sortOrder,
```

Comment:
Is it intensional that there is no cache implemented for certain methods?

Reply:
Added caching for several methods. In resource listing methods we anyway need to to the database search. For some methods like getAPIResources we need to fetch these from the database to preserve that pagination.